File: frontend/src/types/goal.ts
export type GoalStatus = "draft" | "pending" | "approved" | "rejected" | "locked" | "completed" | "on hold"
export type GoalRisk = "Low" | "Medium" | "High"

export interface Milestone {
  id: number
  goal_id: number
  title: string
  due_date: string | null
  is_completed: boolean
  source: "ai" | "manual"
}

export interface Goal {
  id: number
  user_id: number
  title: string
  description: string | null
  target: string | null
  uom: string | null
  weightage: number
  deadline: string | null
  status: GoalStatus
  progress: number
  risk: GoalRisk
  is_shared: boolean
  ai_recommendation: string | null
  created_at: string | null
  updated_at: string | null
  owner_name: string | null
  department: string | null
  milestones: Milestone[]
  task_id?: number | null
  escalations?: Array<{ id: number; goal_id: number; reason: string; severity: string; status: string; admin_remarks?: string; resolution_note?: string; created_at?: string; resolved_at?: string }>
}

export interface GoalCreatePayload {
  title: string
  description?: string
  target?: string
  uom?: string
  weightage: number
  deadline?: string
  task_id?: number
}

export interface GoalUpdatePayload {
  title?: string
  description?: string
  target?: string
  uom?: string
  weightage?: number
  deadline?: string
  progress?: number
}

export interface GoalApprovalPayload {
  action: "approve" | "reject"
  comment?: string
  weightage?: number
  target?: string
}

File: backend/app/models/goal.py
from __future__ import annotations
from datetime import datetime

from sqlalchemy import (
    Boolean,
    DateTime,
    Float,
    ForeignKey,
    Integer,
    String,
    Text,
    func,
)
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.core.database import Base
from app.models.mixins import AuditMixin
from app.models.role import GoalRisk, GoalStatus


class Goal(AuditMixin, Base):
    __tablename__ = "goals"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    user_id: Mapped[int] = mapped_column(Integer, ForeignKey("users.id"), nullable=False, index=True)
    task_id: Mapped[int | None] = mapped_column(Integer, ForeignKey("tasks.id"), nullable=True, index=True)
    title: Mapped[str] = mapped_column(String(255), nullable=False)
    description: Mapped[str | None] = mapped_column(Text, nullable=True)
    target: Mapped[str | None] = mapped_column(String(500), nullable=True)
    uom: Mapped[str | None] = mapped_column(String(50), nullable=True)  # Unit of Measurement
    weightage: Mapped[float] = mapped_column(Float, nullable=False, default=10.0)
    deadline: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)
    status: Mapped[str] = mapped_column(String(20), nullable=False, default=GoalStatus.DRAFT.value, index=True)
    progress: Mapped[float] = mapped_column(Float, default=0.0)
    risk: Mapped[str] = mapped_column(String(10), default=GoalRisk.LOW.value, index=True)
    is_shared: Mapped[bool] = mapped_column(Boolean, default=False, index=True)
    ai_recommendation: Mapped[str | None] = mapped_column(Text, nullable=True)
    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(
        DateTime(timezone=True), server_default=func.now(), onupdate=func.now()
    )

    # Relationships
    owner = relationship("User", back_populates="goals", foreign_keys=[user_id])
    task = relationship("Task", backref="goals")
    milestones = relationship("Milestone", back_populates="goal", cascade="all, delete-orphan", lazy="selectin")
    checkins = relationship("Checkin", back_populates="goal", cascade="all, delete-orphan", lazy="selectin")

    def __repr__(self) -> str:
        return f"<Goal {self.id} '{self.title}' [{self.status}]>"
